### ISSUE TITLE
Author/Endorser: Inconsistent response when publishing SCHEMAs/CRED_DEFs

### ISSUE THREAD

**Issue by frostyfrog** [hyperledger/aries-cloudagent-python#1631]
> Note: Throughout this issue, I use the term "Author" to indicate an Agent with no role on the ledger as part of the Author/Endorser workflow. The term Endorser is used to represent a basic Agent that is not part of the Author/Endorser workflow, but is instead assigned the Endorser role on the ledger.

# The Problem
When a controller is publishing a schema or cred_def through the admin API, the response that they get back differs depending on if they are communicating with an Author or directly to an Endorser. If the running Agent is acting as an Endorser and publishing directly to the ledger, the response is simplified and it's return is delayed until the SCHEMA/CRED_DEF has been published to the ledger. However, if the Agent is running as an Author, the response is returned immediately and follows the documented response, having an object with a `txn` or `sent` property defined (only the `txn` appears to get populated, however).

TL;DR: The output format doesn't match between an Author and an Endorser agent for the same API call.

# Diving into the code

As mentioned earlier, the response format is different from the documentation based upon an agent's role. Let's take a look at the code to determine why.

## POST /schemas
The following is the code that allows a user to publish a new `SCHEMA` to the ledger:
https://github.com/hyperledger/aries-cloudagent-python/blob/70a69d30c559917ea18c30f1504e961d9e3b4d5a/aries_cloudagent/messaging/schemas/routes.py#L153-L289

On line 262, we return a non-schema object upon success. This code point is reached when we communicate with an Endorsing agent directly and instruct it to publish to the ledger. When switching to an Authoring agent, we do not reach this code point.

### Endorser Agent
https://github.com/hyperledger/aries-cloudagent-python/blob/70a69d30c559917ea18c30f1504e961d9e3b4d5a/aries_cloudagent/messaging/schemas/routes.py#L262

### Author Agent
https://github.com/hyperledger/aries-cloudagent-python/blob/70a69d30c559917ea18c30f1504e961d9e3b4d5a/aries_cloudagent/messaging/schemas/routes.py#L289

The Author workflow returns a `TxnOrSchemaSendResultSchema` response, the Endorser workflow does not.

## POST /credential-definitions
The following is the code that allows a user to publish a new `CRED_DEF` to the ledger:
https://github.com/hyperledger/aries-cloudagent-python/blob/70a69d30c559917ea18c30f1504e961d9e3b4d5a/aries_cloudagent/messaging/credential_definitions/routes.py#L151-L304

On line 273, we return a non-schema object upon success. This code point is reached when we communicate with an Endorsing agent directly and instruct it to publish to the ledger. When switching to an Authoring agent, we do not reach this code point.

### Endorser Agent
https://github.com/hyperledger/aries-cloudagent-python/blob/70a69d30c559917ea18c30f1504e961d9e3b4d5a/aries_cloudagent/messaging/credential_definitions/routes.py#L273

### Author Agent
https://github.com/hyperledger/aries-cloudagent-python/blob/70a69d30c559917ea18c30f1504e961d9e3b4d5a/aries_cloudagent/messaging/credential_definitions/routes.py#L304

The Author workflow returns a `TxnOrCredentialDefinitionSendResultSchema` response, the Endorser workflow does not.

**Comment by ianco** [comment]
This is by design.  If the agent only has "author" privileges they can't write to the ledger, and endpoints like "create schema" and "create cred def" can't return the schema and cred def information respectively.  The "transaction" information returned is what is required to get an endorsement signature in order to write the transaction to the ledger.

The controller needs to know, when they call the agent, what privileges the agent has and whether the endorser protocol is required to write to the ledger.

**Comment by frostyfrog** [comment]
I understand that the response has to be different due to the different workflows. Is the difference from the associated `@response_schema` also by design? On a side note (though this probably isn't the best place to ask >_<), what is the recommended workflow for retrieving the `schema_id` and the `credential_definition_id` within the author/endorser workflow?

That's what I was attempting to point out, albeit very poorly as I need to learn how to write better issues in general.
